### PR TITLE
Advanced Search UI

### PR DESCRIPTION
**Pull request by kenkunz**
## Summary

This PR builds on top of #11 with an advanced search page:
* provides a variety of filter and sort options
* displays additional data for each row (price, liquidity, volume)
* shows up to 200 matching results

## Dependencies

This PR depends on #11, which depends on #13.

## Status

I believe this is in good shape to merge. Let me know if you have any questions or feedback on code / user experience.

### Completed

* hide nav search bar on search page
* pass nav-based search query into advanced search
* search page layout (responsive)
* facet-based filters (type, blockchain, exchange)
* numeric range-based filters (volume, liquidity, price change %)
* sort-by options
* clear search options control(s)
* show up to 200 results

### CODE REVIEW

File: src/lib/search/Search.svelte [hunk 1]
```diff
@@ -1,18 +1,22 @@
 <script lang="ts">
   import { goto } from "$app/navigation";
```
Comment (miohtama) [inline src/lib/search/Search.svelte:2]: Add 2-4 lines of comments what this component is about

File: src/routes/search/_FacetFilter.svelte [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<script lang="ts">
+	import { createEventDispatcher } from 'svelte';
```
Comment (miohtama) [inline src/routes/search/_FacetFilter.svelte:2]: Add 2-4 lines of comments what is a facet filter

File: src/routes/search/_Filter.svelte [hunk 3]
```diff
@@ -0,0 +1,54 @@
+<script lang="ts">
+  export let fieldName, options;
```
Comment (miohtama) [inline src/routes/search/_Filter.svelte:2]: Same -add some basic comments on the context of Filter component

File: src/routes/search/_RangeFilter.svelte [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<script lang="ts">
+	import { createEventDispatcher } from 'svelte';
```
Comment (miohtama) [inline src/routes/search/_RangeFilter.svelte:2]: Same - add some context when range filter should be used, what are the inputs needed, etc.

**Review by miohtama — CHANGES_REQUESTED** [review]
Looks very good!

For the future reference, can you add 

- Comment at the top of each component what it does and when it should be used
- One liner comments on component inputs what should go in there

**Comment by kenkunz** [comment]
Used this standard for documenting components:
https://svelte.dev/faq#how-do-i-document-my-components

(see discord thread for more info)

**Comment by kenkunz** [comment]
@miohtama I just push some documentation improvements and a minor refactor. Good to merge from my PoV!

**Review by miohtama — CHANGES_REQUESTED** [review]
There is still some issue with pairs not being sorted correctly (by liquidity).

If I search "MATIC" I get

image

If I use `Liquidity $5M >` filter I get the right pair at the top

image

**Comment by kenkunz** [comment]
@miohtama thanks for the feedback and recommendations!

Re: "advanced search" label in basic search results – 👍 makes sense, will update.

Adding separate comments to address other topics individually.

**Comment by kenkunz** [comment]
## Re: moving the search widget…

Currently on mobile (or anything below 768px) the sort-by widget drops down to the filters area. This is both to accommodate the decreased horizontal space for the search box and to open that slot for the "DONE" button (which closes the expanded filter area).
sort-by-mobile

On wider screens… there are several UX considerations in favor of keeping it in its current location:
* users expect to find it here – it's a very typical / expected location for sort-by widget
* it makes ideal use of available horizontal real-estate
* if we move it somewhere less obvious like the _bottom_ of the filters, I'm concerned it will be buried / lost and most users will miss it

More generally, different users search/sort in different ways (even when seeking similar information). Some users refine results by adding additional search terms; others refine by making use of secondary filters; still others hone in on specific results by sorting on a particular field to "pull" the right results to the top. I'm concerned that moving sort-by to a less obvious location would negatively impact users who's default approach is refinement-by-sort. Keeping it in its current position doesn't negatively impact other strategies.

If you're still in favor of moving it… can I suggest keeping it at the _top_ of the filters? As a drop-down, it has a consistent (small) size relative to the filters, so keeping it at the top has minimal impact on other filters but keeps it accessible for users who care about it.

**Comment by kenkunz** [comment]
## Re: search result order / default sort

Based on tradingstrategy-ai/search#1, the current default sort is `_text_match:desc` (aka relevance) + `volume_24h:desc` (_not_ `liquidity`). Normally… `volume_24h` and `liquidity` would have pretty high correlation … but because of the Polygon issue, a lot of pairs are showing `volume_24h: $0.00` even though they have high liquidity.

> **Side-note:** on basic search, we're also sorting by `type_rank`. This only makes sense in the "basic" context where we show limited search results and use a `group_by:type` to yield _some_ results across different entity types. It doesn't make sense to use on advanced search – you would end up with the highest `type_rank` values dominating the results.

Something to be aware of If we go with something other than `_text_match` (relevance) as the first sort field by default – you will _almost never see exchanges or tokens in search results_ (by default) … b/c the lower relevance _pairs_ will always "win" based on having liquidity and/or volume values (despite lower relevance in text match).

Because of this, my recommendation is:
* rename the default to "relevance" (a more common user-facing term vs. the more technical "text match")
* include `liquidity` AND `volume_24h` as secondary / ternary sort fields when sorting by relevance (useful in scenarios when one of them isn't populated correctly)

Based on your comments above, it sounds like you want `liquidity` to take precedence over `volume_24h`. Is that correct? (Just want to confirm b/c we previously went with `volume_24h`). For comparison… the [Trading pairs index](https://tradingstrategy.ai/trading-view/trading-pairs) page and other index views currently sort by `volume_30d:desc` by default.

There is additional fine-tuning we can do by leveraging other [Typesense search params](https://typesense.org/docs/0.22.2/api/documents.html#search-parameters) – e.g., `query_by_weights` or by playing with the list and order of `query_by` fields. I think this will ultimately provide a better default experience rather than pulling `_text_match` out.

Thoughts?

**Comment by kenkunz** [comment]
## Some additional analysis on search results…

I've been trying to figure out why some of the `MATIC` results score higher on `text_match` than others. In theory they should all match equally, resulting in a tie-break using secondary sort field. Instead, some are matching slightly higher (about `0.04%`).

Here's an example query (note - this is all public data / public-facing search API key):

```
export TYPESENSE_BASE_URL=https://4relmbjhcysqztv9p-1.a1.typesense.net
export TYPESENSE_SEARCH_API_KEY=G26ReHm1VZwTpKye2w4P5hZkmQghb6i9
export JQ_MAPPER='.hits[] | { description:.document.description, tickers:.document.token_tickers|join(", "), names:.document.token_names|join(", "), text_match }'

export Q=MATIC

bash
curl -G "${TYPESENSE_BASE_URL}/collections/trading-entities/documents/search" \
     -H "X-TYPESENSE-API-KEY:${TYPESENSE_SEARCH_API_KEY}" \
     -d "q=${Q}" \
     -d "query_by=description,token_tickers,token_names,smart_contract_addresses" \
     -d "sort_by=_text_match:desc,volume_24h:desc,liquidity:desc" \
     -d "per_page=200" \
| jq $JQ_MAPPER | less
```

Page down through the results until you see the `text_match` value change. Looking at the indexed search fields, I don't see a reason some are matching higher than others. I'll try asking about this in the Typesense slack workspace.

**Comment by miohtama** [comment]
> rename the default to "relevance" (a more common user-facing term vs. the more technical "text match")

There is no need for relevant matches because any match needs to be a substring match in any case. They either enter token name or token symbol in the search box and it needs to be a full match. There is no use case where a user would search fuzzy text results.

> Based on your comments above, it sounds like you want liquidity to take precedence over volume_24h. Is that correct? 

I think this is the way. The current conditions on Polygon network (halted because broken, no volume for 24h) shows that volume is not a good indicator under fault conditions. Whereas liquidity stays in the chain, because no one can do trades, so liquidity cannot change.

So if we switch liquidity > volume_24 I think we have solved this little issue of MATIC token.

**Comment by kenkunz** [comment]
> There is no need for relevant matches because any match needs to be a substring match in any case. They either enter token name or token symbol in the search box and it needs to be a full match. There is no use case where a user would search fuzzy text results.

Good point – I see what you're saying – we're not doing fuzzy searches in the traditional sense. (This will come into play when we index blog posts and documentation.)

There are still some ways in which _relevance_ is still _relevant_, though 😉:
* typos
* instant results as you type – you will match substrings but still get higher relevance value on exact matches
* we are searching across multiple indexed fields with prioritization based on order

To see this in action try a search like "matic ethereum" but pause at "matic eth" and "matic ether" while sorting by `text_match` (relevance). With the full search, the top result is "MATIC on Ethereum" – which seems like the desired result in this instance (you won't see this is the default sort is volume or liquidity). Of course, with this example, we don't really know when the user will stop typing, so we're trying to show them the best results based on what they've given us thus far.

Do you think the label "relevance" is confusing and we should stick with the label "text search"? To me, "relevance" feels right (familiar default sort option) albeit someone vague ("what determines relevance??"). My suggestion would be to use "relevance" as the _short_ label (what users see for the selected option), with an expanded _long_ label for the drop-down option, e.g., "relevance (text match, liquidity)".

**Pull request merged**